python/ray/serve/_private/metrics_utils.py

```diff
@@ -1,138 +1,98 @@
 from dataclasses import dataclass, field
-from typing import Any, Callable, DefaultDict, Dict, List, Optional, Union
+from typing import Callable, DefaultDict, Dict, List, Optional
 
-import ray
 from ray.serve._private.constants import SERVE_LOGGER_NAME
 
 logger = logging.getLogger(SERVE_LOGGER_NAME)
 
 
-class _MetricTask:
-    def __init__(self, task_func, interval_s, callback_func):
-        """
-        Args:
-            task_func: a callable that MetricsPusher will try to call in
-                each loop. It should take no arguments, and return a
-                kwargs dictionary to be used to call the callback_func.
-            interval_s: the interval of each task_func is supposed to be called.
-            callback_func: callback function is called when task_func is done, and
-                the result of task_func is passed to callback_func as the first
-                argument, and the timestamp of the call is passed as the second
-                argument.
-        """
-        self.task_func: Callable[[], Dict[str]] = task_func
-        self.interval_s: float = interval_s
-        self.callback_func: Callable[[Any, float]] = callback_func
-        self.last_ref: Optional[ray.ObjectRef] = None
-        self.last_call_succeeded_time: Optional[float] = time.time()
+@dataclass
+class _MetricsTask:
+    task_func: Callable
+    interval_s: float
 
 
 class MetricsPusher:
-    """
-    Metrics pusher is a background thread that run the registered tasks in a loop.
-    """
+    """Periodically runs registered asyncio tasks."""
+
+    def __init__(
+        self,
+        event_loop: asyncio.AbstractEventLoop,
+        async_sleep: Optional[Callable] = None,
+    ):
+        self._event_loop = event_loop
+        self._async_sleep = async_sleep or asyncio.sleep
+        self._tasks: Dict[str, _MetricsTask] = dict()
+        self._async_tasks: Dict[str, asyncio.Task] = dict()
+        self._stop_event = asyncio.Event(loop=event_loop)
 
-    # If no tasks are registered, sleep for 1s between iterations (until one is
-    # registered).
-    NO_TASKS_REGISTERED_INTERVAL_S = 1
+    def start(self):
+        self._stop_event.clear()
 
-    def __init__(self):
-        self.tasks: Dict[str, _MetricTask] = dict()
-        self.pusher_thread: Union[threading.Thread, None] = None
-        self.stop_event = threading.Event()
+    async def metrics_task(self, name: str):
+        """Periodically runs `task_func` every `interval_s` until `stop_event` is set.
+
+        If `task_func` raises an error, an exception will be logged.
+        """
+
+        while True:
+            if self._stop_event.is_set():
+                return
+
+            try:
+                self._tasks[name].task_func()
+            except Exception as e:
+                logger.exception(f"Failed to run metrics task: {e}")
+
+            _, pending = await asyncio.wait(
+                [
+                    asyncio.create_task(self._stop_event.wait()),
```

Comment:
Should only need to create this task once outside the loop instead of in every iteration

Reply:
Yeah that's what I had originally and switched to this to debug a test failure, sorry - I've switched it back!